src/iperf_api.c

```diff
@@ -4011,3 +4041,252 @@ iflush(struct iperf_test *test)
+        if (thr->stream->sender) {
+            iperf_thread_send(thr);
+        }
+        else {
+            iperf_thread_recv(thr);
+        }
+    }
+    return NULL;
+}
+
+int
+iperf_thread_send(struct iperf_thread *thr)
+{
+    register int multisend, r, streams_active;
+    struct iperf_test *test= thr->test;
+    register struct iperf_stream *sp = thr->stream;
+    struct timeval now;
+
+    /* Can we do multisend mode? */
+    if (test->settings->burst != 0)
+        multisend = test->settings->burst;
+    else if (test->settings->rate == 0)
+        multisend = test->multisend;
+    else
+        multisend = 1;  /* nope */
+
+    for (; multisend > 0; --multisend) {
+        if (test->settings->rate != 0 && test->settings->burst == 0)
+            gettimeofday(&now, NULL);
+        streams_active = 0;
+
+        if (sp->green_light) {
+            if ((r = sp->snd(sp)) < 0) {
+                if (r == NET_SOFTERROR)
+                    break;
+                i_errno = IESTREAMWRITE;
+                return r;
+            }
+            streams_active = 1;
+
+//            pthread_mutex_lock(&thr->test->thrcontrol->send_mutex);
+//            test->bytes_sent += r;
+//            ++test->blocks_sent;
+//            pthread_mutex_unlock(&thr->test->thrcontrol->send_mutex);
+
+            sp->bytes_sent += r;
+            ++sp->blocks_sent;
+
+            if (test->settings->rate != 0 && test->settings->burst == 0)
+                iperf_check_throttle(sp, &now);
+            if (multisend > 1 && test->settings->bytes != 0 && test->bytes_sent >= test->settings->bytes)
+                break;
+            if (multisend > 1 && test->settings->blocks != 0 && test->blocks_sent >= test->settings->blocks)
+                break;
+        }
+
+        if (!streams_active)
+            break;
+    }
+
+    if (test->settings->burst != 0) {
+        gettimeofday(&now, NULL);
+        iperf_check_throttle(sp, &now);
+    }
+
+    return 0;
+}
+
+int
+iperf_thread_recv(struct iperf_thread *thr)
+{
+    int r;
+
+    if ((r = thr->stream->rcv(thr->stream)) < 0) {
+        i_errno = IESTREAMREAD;
+        return r;
+    }
+
+    /* XXX: bytes_received and blocks_received are never used. Do we need to count it? */
+//    pthread_mutex_lock(&thr->test->thrcontrol->receive_mutex);
+//    thr->test->bytes_received += r;
+//    ++thr->test->blocks_received;
+//    pthread_mutex_unlock(&thr->test->thrcontrol->receive_mutex);
+
+    return 0;
+}
+
+int
+iperf_delete_threads(struct iperf_test *test)
+{
+    struct iperf_thread *thr;
+    struct iperf_threads_control *control;
+    int i;
+
+    control = test->thrcontrol;
+
+    if (control) {
+        for (i = 0; i < test->thrcontrol->sum_threads; ++i) {
+            thr = test->thrcontrol->threads[i];
+            if (thr) {
+                pthread_join(thr->thread, NULL);
+                free(thr);
+                thr = NULL;
+            }
+        }
+
+        if (control->threads) {
```

Comment:
Must check validity of  `control->threads` before the loop in which it is accessed.